File: 2017_grid_game.java
class Solution {  
    public long gridGame(int[][] grid) {  
        int n = grid[0].length;  
        long[] top = new long[n + 1], bot = new long[n + 1];  
        for (int i = 0; i < n; i++) {  
            top[i + 1] = top[i] + grid[0][i];  
            bot[i + 1] = bot[i] + grid[1][i];  
        }  
        long minPoints = Long.MAX_VALUE;  
        for (int i = 0; i < n; i++)  
            minPoints = Math.min(minPoints, Math.max(top[n] - top[i + 1], bot[i]));  
        return minPoints;  
    }  
}  

File: 2017_grid_game.cpp
class Solution {  
public:  
    long long gridGame(vector<vector<int>>& g) {  
        int n = g[0].size();  
        vector<long long> p1(n + 1), p2(n + 1);  
        for (int i = 0; i < n; ++i) {  
            p1[i + 1] = p1[i] + g[0][i];  
            p2[i + 1] = p2[i] + g[1][i];  
        }  
        long long mn = LLONG_MAX;  
        for (int i = 0; i < n; ++i)  
            mn = min(mn, max(p1[n] - p1[i + 1], p2[i]));  
        return mn;  
    }  
};  
